pkg/resource/deploy/step_generator.go

```diff
@@ -2850,18 +2851,53 @@ func (sg *stepGenerator) AnalyzeResources() error {
 				Dependencies:         v.Dependencies,
 				PropertyDependencies: v.PropertyDependencies,
 			}
-			providerResource := sg.getProviderResource(v.URN, v.Provider)
+			// N.B. This feels very unideal but I can't find a better way to check this. When we get here
+			// there is a chance that we'll have a resource in `news` but _won't_ have it's matching provider.
+			// This can happen in a targeted run where the untargeted resource is removed from the program
+			// and is the only resource using the given provider (common case is where every other resource is
+			// using a new version of the provider).
+			// See https://github.com/pulumi/pulumi/issues/19879 for a case of this.
+			var providerResource *resource.State
+			if v.Provider != "" {
+				var ref providers.Reference
+				ref, err = providers.ParseReference(v.Provider)
+				contract.AssertNoErrorf(err, "failed to parse provider reference")
+				var has bool
+				providerResource, has = sg.providers[ref.URN()]
+				if !has {
+					// This provider hasn't been registered yet. This happens when a user changes the default
+					// provider version in a targeted update. See https://github.com/pulumi/pulumi/issues/15732
+					// for more information.
+					for _, r := range sg.deployment.olds {
```

Comment:
Would it work if you did `sg.deployment.olds[ref.URN()]` and then checked the ID? Saves you the loop over all the resources if so, which feels a bit more performant? Also not clear to me when we prefer `.Olds()` vs `.olds` but that's neither here nor there.